java/src/jmri/jmrix/mqtt/MqttAdapter.java

```diff
@@ -61,7 +61,7 @@ public void connect() throws IOException {
                 options.put(option2Name, new Option("MQTT channel :", new String[]{baseTopic, DEFAULT_BASETOPIC}));
             }
 
-            String clientID = CLID + "-" + this.getUserName();
+            String clientID = CLID + "-" + this.getUserName() + "-" + jmri.profile.ProfileManager.getDefault().getActiveProfile().getUniqueId();
```

Comment:
This will not ensure sufficient uniqueness, since if two JMRI instances are running the same profile, they will use the same profile ID. (The JMRI profile ID system is designed to allowed for this.)

You want to use `jmri.util.node.NodeIdentity.networkIdentity()` instead of `jmri.profile.ProfileManager.getDefault().getActiveProfile().getUniqueId()`.

Reply:
(This change will also resolve the static analysis failures.)

Reply:
networkIdentity() returns a 26 character string on my system, from a quick look at the source code I am assuming this is a fixed length on all systems, the maximum client ID length allowed by the MQTT protocol is 23 characters.

The networkIdentity()  ID begins with "jmri-" if this prefix were excluded then networkIdentity() could be used as the Client ID, which from my reading of the source code for network identity would result in essentially the concatenation of the mac address and profile ID which should ensure uniqueness.

Reply:
The network identity is "jmri-" + MAC address (12 characters) + "-" + profile ID (8 characters).

http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718031 has _minimum_ requirements for an acceptable client ID, which also exclude the use of dashes (as well as minimally allow 23 characters in the ID). An MQTT server may allow a longer ID and other characters than those specified.

So your proposal (and what JMRI had) is also not ensured to be valid name.

To make a safe name that is ensured to be valid, I would use:
```
String clientID = (NodeIdentity.networkIdentity() + getSystemPrefix()).replaceAll("/[^A-Za-z0-9 ]/", "");
if (clientID.length() > 23) {
    clientID = clientID.subString(clientID.length() - 23);
}
```

Note that I am truncating from the left, not the right, in an attempt to allow:
- Single JMRI instance to have 2 or more MQTT connections to a single MQTT server
- Single computer to run multiple JMRI instances with connections to a single MQTT server (assuming each instance is using different profile)
- Multiple computers running a JMRI instance triggered from copies of the same profile to a single MQTT server

Reply:
This works with the following modified regex (removed leading and trailing /)  "[^A-Za-z0-9 ]" as it did not remove the -‘s with them in.